fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/DataCachePopulateRewriteRule.java

```diff
@@ -0,0 +1,184 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.tree;
+
+import com.starrocks.catalog.Table;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.datacache.DataCacheOptions;
+import com.starrocks.datacache.DataCachePopulateMode;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.qe.SessionVariable;
+import com.starrocks.sql.ast.QueryStatement;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.ScanOperatorPredicates;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalScanOperator;
+import com.starrocks.sql.optimizer.task.TaskContext;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+public class DataCachePopulateRewriteRule implements TreeRewriteRule {
+    private static final Logger LOG = LogManager.getLogger(DataCachePopulateRewriteRule.class);
+
+    private final ConnectContext connectContext;
+
+    public DataCachePopulateRewriteRule(ConnectContext connectContext) {
+        this.connectContext = connectContext;
+    }
+
+    @Override
+    public OptExpression rewrite(OptExpression root, TaskContext taskContext) {
+        SessionVariable sessionVariable = connectContext.getSessionVariable();
+        if (!sessionVariable.isEnableScanDataCache()) {
+            return root;
+        }
+        DataCachePopulateMode populateMode = getPopulateMode(sessionVariable);
+
+        DataCachePopulateRewriteVisitor visitor = new DataCachePopulateRewriteVisitor(populateMode);
+        root.getOp().accept(visitor, root, taskContext);
+        return root;
+    }
+
+    private DataCachePopulateMode getPopulateMode(SessionVariable sessionVariable) {
+        DataCachePopulateMode populateMode = sessionVariable.getDataCachePopulateMode();
+        if (!sessionVariable.isEnablePopulateDataCache()) {
+            // be compatible with old parameter
+            populateMode = DataCachePopulateMode.NEVER;
+        }
+
+        // if populateMode is auto, we have to check further
+        if (populateMode == DataCachePopulateMode.AUTO) {
+            if (!isSatisfiedStatement()) {
+                populateMode = DataCachePopulateMode.NEVER;
+            }
+        }
+        return populateMode;
+    }
+
+    private boolean isSatisfiedStatement() {
+        // check is analyze sql
+        if (connectContext.isStatisticsJob()) {
+            return false;
+        }
+        // check is query statement
+        // make sure executor is not nullptr first
+        if (connectContext.getExecutor() == null) {
+            return false;
+        }
+        if (!(connectContext.getExecutor().getParsedStmt() instanceof QueryStatement)) {
```

Comment:
Will `cache select` be affected by it ?

Reply:
fix it, set populate mode in CacheSelectExecutor.